### PR TITLE
hoa router

### PR DESCRIPTION
**Pull request by speier**
router implemented as a higher order app (hoa)

### CODE REVIEW

File: src/router.js [hunk 1]
```diff
@@ -1,41 +1,38 @@
...
+              actions.set({ path: path })
             }
           }
         }
-      },
-      events: {
-        load: function(state, actions) {
-          addEventListener("popstate", function() {
+
+        props.view = function (state, actions) {
```
Comment (pspeter3) [inline src/router.js]: This seems like a potentially dangerous pattern with HOA. What happens if a view was defined in the existing app chain?

File: src/router.js [hunk 2]
```diff
@@ -1,41 +1,38 @@
...
+          window.onpopstate = function () {
```
Comment (pspeter3) [inline src/router.js]: Can this be an event listener as opposed to a handler? Overriding global state again seems concerning. Also why does this happen in view as opposed to looking at the actions returned by the call to app?

Comment (speier) [reply src/router.js]: unfortunately it wasn't working outside of the view. i prefer to use `addEventListener` outside the view but can't make it working.

File: src/router.js (hunk 1 above)
Comment (speier) [reply src/router.js]: events gone, so the only possibly entrypoint is the view function, but i think that is ok. the previous implementation of the router was doing pretty much the same on the `render` event.

Comment (speier) [reply src/router.js]: please see my lates commit, that is overwriting the default view only when the route matches

File: src/router.js (hunk 2 above)
Comment (speier) [reply src/router.js]: please see may improved implementation in the latest commit; `addEventListener` working

File: src/router.js [hunk 3]
```diff
@@ -1,41 +1,41 @@
...
+      const actions = app(enhance(props))
+
+      window.addEventListener('popstate', function () {
+        actions.router.set({})
+      })
+
+      return actions
+
+      function enhance(props) {
+        props = Object.assign({ state: {}, actions: {} }, props)
```
Comment (SkaterDad) [inline src/router.js]: hyperapp's core aims to support IE10+, so you may need to replace `Object.assign` with a custom function to maintain that compatibility.

Maybe @JorgeBucaran would consider exporting the `merge()` function from hyperapp/src/app.js so other packages can use it and avoid duplicating efforts/code?

https://github.com/hyperapp/hyperapp/blob/master/src/app.js#L120-L129

It's also possible that his latest PR to introduce modules will mean you don't need this section of code.

Comment (speier) [reply src/router.js]: ok, i will fix that. exporting `merge` would be great. in the meantime how about simply?
```
props = props || {}
props.state = props.state || {}
props.actions = props.actions || {}
```

also, i will check out modules...

**Comment by Swizz** [comment]
Using module we do not need HOA anymore :tada:

File: README.md [hunk 4]
```diff
@@ -6,30 +6,44 @@
...
+    (state, actions) =>
+      <Link to="/" go={actions.router.go}>
+        Back
+      </Link>
+  ]
+]
+
+const routable = router(routes)
+
+app(routable)({
```
Comment (okwolf) [inline README.md]: Support for passing HOAs to `app` has been removed in the latest Hyperapp

File: package.json [hunk 5]
```diff
@@ -37,13 +37,13 @@
   },
   "devDependencies": {
     "babel-preset-es2015": "^6.24.1",
-    "hyperapp": "^0.12.0",
-    "jest": "^20.0.4",
-    "prettier": "~1.5.3",
-    "rollup": "^0.47.6",
+    "hyperapp": "^0.14.0",
```
Comment (okwolf) [inline package.json]: Hyperapp 0.15.1 is out now

**Pull request closed without merging**